### PR TITLE
SDL 0153 Support Short and Full UUID App IDs

### LINKED ISSUE #942: [SDL 0153] Support for Short and Full UUID App ID
Proposal: [Support for Short and Full UUID App ID](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0153-support-short-long-appid.md)

> Some OEM legacy head units can't use an app id that has been generated by [smartdevicelink.com](https://www.smartdevicelink.com). We would like to propose a shortened version of the UUID to be used in the legacy app id field and we add another variable for "fullAppID".

Review: https://github.com/smartdevicelink/sdl_evolution/issues/443

### PR DESCRIPTION
**Pull request by NicoleYarroch**
Fixes #942, #1063

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Test cases added to the _SDLRegisterAppInterface_ and the _SDLLifecycleConfiguration_ spec files

### Summary
Add a new variable, `fullAppID` to the `RegisterAppInterface` request.

##### Enhancements
* Add a new variable, `fullAppID` to the `RegisterAppInterface` request.
* Moved the `SDLLifecycleConfigurationUpdate` test cases to their own spec file.

##### Bug Fixes
* Fixed broken test cases in the _SDLStreamingVideoLifecycleManagerSpec_

### Tasks Remaining:
- [x] Test with Core v.5.0
- [x] Test with Manticore (Core v.4.5.1) 
- [x] Test with SYNC_3

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLRegisterAppInterface.m [hunk 1]
```diff
@@ -42,23 +47,40 @@ - (instancetype)initWithLifecycleConfiguration:(SDLLifecycleConfiguration *)life
                 nightColorScheme:lifecycleConfiguration.nightColorScheme];
 }
 
+- (instancetype)initWithAppName:(NSString *)appName appId:(NSString *)appId fullAppId:(nullable NSString *)fullAppId languageDesired:(SDLLanguage)languageDesired {
+    self = [self initWithAppName:appName languageDesired:languageDesired];
+    if (!self) {
+        return nil;
+    }
+
+    self.fullAppID = fullAppId.length > 0 ? fullAppId : nil;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLRegisterAppInterface.m]: This is confusing to me. Why check `length = 0` and set `nil`. If it's `nil`, then the `length == 0` in Obj-C.

File: SmartDeviceLink/SDLRegisterAppInterface.m [hunk 2]
```diff
@@ -42,23 +47,40 @@ - (instancetype)initWithLifecycleConfiguration:(SDLLifecycleConfiguration *)life
...
+    self.appID = fullAppId.length > 0 ? [self.class sdlex_shortAppIdFromFullAppId:fullAppId] : appId;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLRegisterAppInterface.m]: This should be done in the `SDLLifecycleConfiguration` in a separate private function.

File: SmartDeviceLink/SDLRegisterAppInterface.m [hunk 3]
```diff
@@ -42,23 +47,40 @@ - (instancetype)initWithLifecycleConfiguration:(SDLLifecycleConfiguration *)life
...
 - (instancetype)initWithAppName:(NSString *)appName appId:(NSString *)appId languageDesired:(SDLLanguage)languageDesired {
-    self = [self init];
+    self = [self initWithAppName:appName languageDesired:languageDesired];
     if (!self) {
         return nil;
     }
-    
-    self.appName = appName;
+
+    self.fullAppID = @"";
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLRegisterAppInterface.m]: Why not set this to `nil`?

File: SmartDeviceLink/SDLRegisterAppInterface.m [hunk 4]
```diff
@@ -103,6 +125,29 @@ - (instancetype)initWithAppName:(NSString *)appName appId:(NSString *)appId lang
     return self;
 }
 
+#pragma mark Initializer Helpers
+
+- (instancetype)initWithAppName:(NSString *)appName languageDesired:(SDLLanguage)languageDesired {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLRegisterAppInterface.m]: This doesn't have a public corollary. It doesn't need to exist. All of the other initializers should be able to call the one initializer with all properties I think. All of these initializer implementations need some cleanup.

File: SmartDeviceLink/SDLLifecycleConfiguration.h [hunk 5]
```diff
@@ -69,9 +91,18 @@ NS_ASSUME_NONNULL_BEGIN
 
 /**
  *  The app id. This must be the same as the app id received from the SDL developer portal or OEM.
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLLifecycleConfiguration.h]: I added more documentation. Right now the `appId` will only be automatically set if one of the convenience inits is used to create the `SDLLifecycleConfiguration` object.

File: SmartDeviceLink/SDLLifecycleConfiguration.h [hunk 6]
```diff
@@ -69,9 +91,18 @@ NS_ASSUME_NONNULL_BEGIN
...
+ *
+ *  Required
  */
 @property (copy, nonatomic, readonly) NSString *appId;
 
+/**
+ * The app id. This must be the same as the app id received from the SDL developer portal or OEM.
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLLifecycleConfiguration.h]: Fixed

File: SmartDeviceLink/SDLLifecycleConfiguration.m [hunk 7]
```diff
@@ -50,23 +66,22 @@ - (instancetype)initDefaultConfigurationWithAppName:(NSString *)appName appId:(N
     _ttsName = nil;
     _voiceRecognitionCommandNames = nil;
 
-    return self;
-}
+    _fullAppId = fullAppId;
+    _appId = appId;
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLLifecycleConfiguration.m]: Fixed

File: SmartDeviceLink/SDLLifecycleConfiguration.m [hunk 8]
```diff
@@ -50,23 +66,22 @@ - (instancetype)initDefaultConfigurationWithAppName:(NSString *)appName appId:(N
...
 
-+ (SDLLifecycleConfiguration *)defaultConfigurationWithAppName:(NSString *)appName appId:(NSString *)appId {
-    return [[self alloc] initDefaultConfigurationWithAppName:appName appId:appId];
+    return self;
 }
 
-+ (SDLLifecycleConfiguration *)debugConfigurationWithAppName:(NSString *)appName appId:(NSString *)appId ipAddress:(NSString *)ipAddress port:(UInt16)port {
-    SDLLifecycleConfiguration *config = [[self alloc] initDefaultConfigurationWithAppName:appName appId:appId];
++ (SDLLifecycleConfiguration *)debugConfigurationWithAppName:(NSString *)appName fullAppId:(NSString *)fullAppId appId:(NSString *)appId ipAddress:(NSString *)ipAddress port:(UInt16)port {
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLLifecycleConfiguration.m]: Changed. This was a helper function.

File: SmartDeviceLink/SDLLifecycleConfiguration.m [hunk 9]
```diff
@@ -29,7 +29,25 @@ @implementation SDLLifecycleConfiguration
...
+    return [self debugConfigurationWithAppName:appName fullAppId:@"" appId:appId ipAddress:ipAddress port:port];
+}
+
++ (SDLLifecycleConfiguration *)debugConfigurationWithAppName:(NSString *)appName fullAppId:(NSString *)fullAppId ipAddress:(NSString *)ipAddress port:(UInt16)port {
+    return [self debugConfigurationWithAppName:appName fullAppId:fullAppId appId:@"" ipAddress:ipAddress port:port];
+}
+
+#pragma mark Initalization Helpers
+
+- (instancetype)initDefaultConfigurationWithAppName:(NSString *)appName fullAppId:(NSString *)fullAppId appId:(NSString *)appId  {
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLLifecycleConfiguration.m]: Changed.

File: SmartDeviceLink/SDLRegisterAppInterface.m (hunk 4 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLRegisterAppInterface.m]: Removed

File: SmartDeviceLink/SDLLifecycleConfiguration.m [hunk 10]
```diff
@@ -93,11 +110,36 @@ - (void)setAppType:(nullable SDLAppHMIType)appType {
...
+ *  Generates the `appId` from the `fullAppId`
+ *
+ *  @discussion When an app is registered with an OEM, it is assigned an `appID` and a `fullAppID`. The `fullAppID` is the full UUID appID. The `appID` is the first 10 non-dash (i.e. "-") characters of the  `fullAppID`.
+ *
+ *  @param fullAppId   A `fullAppId`
+ *  @return            An `appID` made of the first 10 non-dash characters of the "fullAppID"
+ */
++ (NSString *)sdlex_shortAppIdFromFullAppId:(NSString *)fullAppId {
+    NSString *filteredString = [self sdlex_filterDashesFromText:fullAppId];
+    return [filteredString substringToIndex:MIN(AppIdCharacterCount, filteredString.length)];
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLLifecycleConfiguration.m]: It won't crash as the code uses either 10 or the length of the filtered string, whichever is smaller, to create the substring. There are test cases for this situation in the _SDLLifecycleConfigurationSpec_

File: SmartDeviceLink/SDLRegisterAppInterface.h [hunk 11]
```diff
@@ -14,90 +14,96 @@
...
 NS_ASSUME_NONNULL_BEGIN
 
 @interface SDLRegisterAppInterface : SDLRPCRequest
 
+/**
+ * Init
+ *
+ * @return A SDLRegisterAppInterface object
+ */
+- (instancetype)init;
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLRegisterAppInterface.h]: Fixed. Removed

**Pull request merged**